### PR TITLE
first attempt at dynamic HUD widget positioning

### PR DESCRIPTION
**Pull request by fabiangreffrath**
This is a first sketch of what I'd like to achieve with dynamic HUD widget positioning. It is still a bit rough around the edges but works in general.

What's still missing:

- [ ] Polish. :wink: 
- [x] Proper right-alignment. Currently all `topright` and `bottomright` widgets are simply drawn at `x = 200 + WIDESCREENDELTA` coordinate.
- [x] Replacement of the `w_lstat?` widgets by `w_monsec`, and `w_ltime` by `w_sttime`.
- [x] Condensation of the `w_coord?` widgets into a single `w_coord` widget.
- [x] Optional use of bold font for all widgets.
- [x] Filling of the `widgets[]` array with contents of the `WOOFHUD` lump.
- [x] Sometimes, with uncapped framerate enabled, you can see for a split-second how the widgets rearrange when changing views.
- [x] The widget initialization sequence in HU_Start() could be vastly simplified, now that widget coordinates are determined at drawing time.
- [x] Menu option consistence: For the Automap, we have two separate on/off toggles for map stats and map time. For the Status Bar / Crispy HUD, we have a combined off/stats/time/both toggle. And finally, for the Boom HUD we have a toggle not to hide the stats widget. I'd prefer the solution to have separate on/off toggles for each scenario.
- [ ] and possibly more...

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
>     * and possibly more...

Sometimes, with uncapped framerate enabled, you can see for a split-second how the widgets rearrange when changing views.

**Comment by rfomin** [comment]
Why you didn't merge #886 first?
The level title is always below the extended HUD line:
image
Or is it too early to report problems?

**Comment by fabiangreffrath** [comment]
> Why you didn't merge #886 first? 

Waiting for your respected opinion. :wink: 

> The level title is always below the extended HUD line: 

Yes, because that's where Vanilla prints it, right?

**Comment by rfomin** [comment]
> Yes, because that's where Vanilla prints it, right?

In the previous version, we did it differently, but this is a matter of taste.

**Comment by fabiangreffrath** [comment]
Two more things to consider:
* The widget initialization sequence in `HU_Start()` could be vastly simplified, now that widget coordinates are determined at drawing time.
* Menu option consistence: For the Automap, we have two separate on/off toggles for map stats and map time. For the Status Bar / Crispy HUD, we have a combined off/stats/time/both toggle. And finally, for the Boom HUD we have a toggle *not to hide* the stats widget. I'd prefer the solution to have separate on/off toggles for each scenario.

**Comment by fabiangreffrath** [comment]
@rfomin How do you like the concept in general, to decide which widgets are drawn by constructing them in the ticker function, and to decide in which order and where exactly they are drawn in the drawer function?

**Comment by fabiangreffrath** [comment]
> Sounds reasonable. My main concern is not getting convoluted code due to multiple hud modes.

Yes. Currently, the most convulted part is in / has been moved to `HU_Ticker()` where each of the possible HUD states is checked for. We could possibly get a bit clearer here, but over all, I think the distinction is quite reasonable. Suggestions are always welcome!

File: src/hu_lib.c [hunk 1]
```diff
@@ -437,9 +490,9 @@ void HUlib_addMessageToMText(hu_mtext_t *m, char *prefix, char *msg)
 // Passed a hu_mtext_t
 // Returns nothing
 //
-// killough 11/98: Simplified, allowed text to scroll in either direction
+// killough 11/98: Simplified, allowed text to scroll in either align_direction
```
Comment (fabiangreffrath) [inline src/hu_lib.c:493]: Oops!

File: src/hu_stuff.c [hunk 2]
```diff
@@ -1168,11 +1168,11 @@ static void HU_widget_build_monsec(void)
     }
   }
 
-  kills_color = (kills - extrakills >= totalkills) ? '0'+CR_BLUE : '0'+CR_GOLD;
-  kills_percent_color = (kills >= totalkills) ? '0'+CR_BLUE : '0'+CR_GOLD;
+  kills_color = (kills - extrakills >= totalkills) ? '0'+CR_BLUE : '0'+CR_GRAY;
```
Comment (fabiangreffrath) [inline src/hu_stuff.c:1173]: I couldn't stand the all-yellow font anymore. What do you say @rfomin ?

Comment (rfomin) [reply src/hu_stuff.c:1173]: I assume it's for consistency with vanilla:
image
Although, I have no preference for font colors.

**Comment by fabiangreffrath** [comment]
Fixes #593

File: src/m_menu.c [hunk 3]
```diff
@@ -3297,7 +3297,8 @@ setup_menu_t stat_settings1[] =  // Status Bar and HUD Settings screen
 
   {"HEADS-UP DISPLAY"  ,S_SKIP|S_TITLE,m_null,M_X,M_Y+stat1_title2*M_SPC},
 
-  {"HIDE LEVEL STATS"  ,S_YESNO|S_COSMETIC,m_null,M_X,M_Y+stat1_stats*M_SPC, {"hud_nosecrets"}},
+  {"SHOW LEVEL STATS"  ,S_YESNO|S_COSMETIC,m_null,M_X,M_Y+stat1_stats*M_SPC, {"hud_level_stats"}},
+  {"SHOW LEVEL TIME"   ,S_YESNO|S_COSMETIC,m_null,M_X,M_Y+stat1_time*M_SPC,  {"hud_level_time"}},
```
Comment (rfomin) [inline src/m_menu.c:3301]: What if we somehow highlight these settings? For example, let's move the "Extended HUD" group to the first page and add them to it. I have been looking for a long time where my favorite option has gone 😄

**Review by rfomin — COMMENTED** [review]
Overall this looks great, thank you!

A few questions/suggestions about usability:
1. What do you think if we set the automap widgets to a "large" font by default? Also I think we need a separate option for this, it's confusing that the settings in the HUD menu affect the automap widgets.
2. I think we should move player coordinates to topright or topleft, often it's not fit on screen:
woof0000
3. I'm confused about how distributed mode works now, maybe we should simplify or just get rid of it. Although I like the distributed Crispy HUD.

**Comment by rfomin** [comment]
Also I suggest to just enable the FPS widget all the time if it is present in `woofhud.lmp`.

**Comment by fabiangreffrath** [comment]
> Also I suggest to just enable the FPS widget all the time if it is present in `woofhud.lmp`.

The point is it will never show if it's not present in the lump. Or more general, if you define a HUD in that lump and leave one widget out, it will *never* show up. The cheat code is just there to trigger its presence, but it needs to be positioned in the lump in any case.

**Comment by fabiangreffrath** [comment]
> Overall this looks great, thank you!

😎 

> 1. What do you think if we set the automap widgets to a "large" font by default? Also I think we need a separate option for this, it's confusing that the settings in the HUD menu affect the automap widgets.

Sure, this can be done. We can change the `hu_font2` pointer depending on `automapactive`.

> 2. I think we should move player coordinates to topright or topleft, often it's not fit on screen:

Alright. In MBF they were in the topright corner, so let's move them back there.

> 3. I'm confused about how distributed mode works now, maybe we should simplify or just get rid of it. Although I like the distributed Crispy HUD.

It works as before. The widgets are slammed into the screen corners. I find it a bit confusing that the keys/monsec/time widgets need another F5 to be shown, but it was like that before, or not?

File: src/hu_stuff.c [hunk 4]
```diff
@@ -1534,123 +1347,42 @@ void HU_Drawer(void)
...
   }
 
+  HUlib_resetAlignOffsets();
+
+  if (message_list)
+    HUlib_drawMText(&w_rtext, align_text);
+  else
+    HUlib_drawSText(&w_message, align_text);
+
+  HUlib_drawSText(&w_secret, align_text);
```
Comment (rfomin) [inline src/hu_stuff.c:1357]: Should be `align_direct` for the secret message, right?

Comment (fabiangreffrath) [reply src/hu_stuff.c:1357]: Pretty sure yes!

**Comment by rfomin** [comment]
> but it was like that before, or not?

I think something has changed from the master. Try to set the Crispy HUD + woofhud.lmp from your example and press F5 a few times in full screen mode.

**Comment by fabiangreffrath** [comment]
Yes. Distributed Crispy HUD corresponds to "hud 1", so the monsec+time widgets get moved to the upper right corner.

**Comment by rfomin** [comment]
> Yes. Distributed Crispy HUD corresponds to "hud 1", so the monsec+time widgets get moved to the upper right corner.

I think it's confusing, I still don't understand how it works. I propose to somehow simplify the distributed mode or just remove it. I used to have a different opinion, sorry for that. I think users will also be confused.

**Comment by fabiangreffrath** [comment]
> I think it's confusing, I still don't understand how it works. I propose to somehow simplify the distributed mode or just remove it.

Hm, sure we could do that and just iterate through the available HUDs.

But how about the (distributed) Crispy HUD then? On the master branch we have hard-coded the positions of the monsec+time widgets to be bottomleft right above the status bar widgets. Now they would have dynamic positions which correspond to the ones parsed from the WOOFHUD lump. Does this feel counter intuitive?

**Comment by rfomin** [comment]
I think I've found the problem, I also have the player coords automap widget set to "always". If you press F5 with the Crispy HUD + woofhud.lmp and automap coords enabled, something weird will definitely happen.

**Comment by fabiangreffrath** [comment]
> I think it's confusing, I still don't understand how it works.

Does this help?
```patch
diff --git a/src/hu_stuff.c b/src/hu_stuff.c
index 632ff93..1e120f5 100644
--- a/src/hu_stuff.c
+++ b/src/hu_stuff.c
@@ -1614,7 +1614,6 @@ void HU_Ticker(void)
       HU_widget_build_health();
       HU_widget_build_ammo();
 
-      if (hud_active > 1)
       {
         HU_widget_build_keys();
 
diff --git a/src/m_menu.c b/src/m_menu.c
index aca2127..4ae3a9b 100644
--- a/src/m_menu.c
+++ b/src/m_menu.c
@@ -5547,7 +5547,7 @@ boolean M_Responder (event_t* ev)
          else
            {
              hud_displayed = 1;               //jff 3/3/98 turn hud on
-             hud_active = crispy_hud ? !hud_active : (hud_active+1)%3; // cycle hud_active
+             hud_active = !hud_active;
              if (!hud_active)                 //jff 3/4/98 add distributed
                {
                  hud_distributed = !hud_distributed; // to cycle
```

**Comment by fabiangreffrath** [comment]
> If you press F5 with the Crispy HUD + woofhud.lmp and automap coords enabled, something weird will definitely happen.

I don't think so. The coord widget changes its position between the two different HUDs, i.e. topright in compact mode and topleft in distributed mode. It also does so if the other widgets are not shown, i.e. in the "gaps" between compact and distributed mode, because it is set to be "always" visible.

**Comment by rfomin** [comment]
> Does this help?

No, still confused. I press F5:
1. 
woof0001
2.
woof0002
3.
woof0003
4.
woof0004
Was that how it was intended?

**Comment by fabiangreffrath** [comment]
1. HUD distributed (hud 1) => coords topleft,  HUD active = 0 => no other widgets visible
2. HUD distributed (hud 1) => coords topleft,  HUD active = 1 => other widgets visible
3. HUD compact     (hud 0) => coords topright, HUD active = 0 => no other widgets visible
4. HUD compact     (hud 0) => coords topright, HUD active = 1 => other widgets visible

I don't say it was intended, but it appears logical to me. But, granted, I have been "living" in this code for the past 3+ weeks. I have been literally dreaming about this, so maybe me perception is a bit biased. 🤣

**Comment by rfomin** [comment]
This is definitely too complicated for me 😄 I suggest ignoring the `WOOFHUD` lump in Crispy mode, so we have only 2 modes - distributed and compact, as it was before.
Perhaps in full screen mode, we need to somehow simplify too? It seems that before we had 2 modes, and now 4.

**Comment by fabiangreffrath** [comment]
> Edit: Oh, before in Boom mode, we also had all these combinations, I just didn’t notice. Maybe it's for another PR then.

🤣 Yes, it's confusing, even before my changes.

In the long term, we could do it like PrBoom+ and just cycle through all the HUDs defined in the WOOFHUD lump (in theory there could be more than just 2), but only ship the basic 2 HUDs (compact and distributed) by default.

Still not sure how to arrange the extra widgets when the Crispy HUD is enabled, though. I wouldn
T want to hard-code their positions - that's where we came from.

**Comment by rfomin** [comment]
> In the long term, we could do it like PrBoom+ and just cycle through all the HUDs defined in the WOOFHUD lump (in theory there could be more than just 2), but only ship the basic 2 HUDs (compact and distributed) by default.

Yes, I like this idea.

> Still not sure how to arrange the extra widgets when the Crispy HUD is enabled, though. I wouldn't want to hard-code their positions - that's where we came from.

I think the Crispy HUD should work the same as the STBAR mode. Although, I like the distributed option. Perhaps hardcoded positions are the most reasonable approach if we don't want to add "big num/icons" widgets.

**Comment by rfomin** [comment]
> @rfomin Is it a bit more intuitive after the latest commit?

Yes, it looks better. Now it seems like 3 modes, not 4, right? I still don't get it, but it looks reasonable.

**Comment by fabiangreffrath** [comment]
> > @rfomin Is it a bit more intuitive after the latest commit?
> 
> Yes, it looks better. Now it seems like 3 modes, not 4, right? I still don't get it, but it looks reasonable.

Lol, right. Now it's off -> compact -> distributed -> (repeat).

**Comment by fabiangreffrath** [comment]
Yes, with the latest commit, the variables and values have changed meanings: HUD 0 is empty, HUD 1 is compact and HUD 2 is distributed. There is no further distinction anymore. 

You may customize any of these HUDs (even the empty one) through the WOOFHUD lump by choosing the corresponding number in the section heading.

**Comment by rfomin** [comment]
Another nitpick - `hud 0` in `WOOFHUD` with Crispy mode enabled:
woof0000

**Comment by fabiangreffrath** [comment]
> Another nitpick - `hud 0` in `WOOFHUD` with Crispy mode enabled

Erm, this should be fixed now, right?

**Comment by rfomin** [comment]
> > Another nitpick - `hud 0` in `WOOFHUD` with Crispy mode enabled
> 
> Erm, this should be fixed now, right?

Yes, but `WOOFHUD` only works in Boom mode, which I think is fine.

**Comment by fabiangreffrath** [comment]
> Yes, but `WOOFHUD` only works in Boom mode, which I think is fine.

Well, not strictly. You can still modify the position of the monsec/sttime/coord/fps widgets in compact and distributed Crispy HUD mode.

**Comment by rfomin** [comment]
> You can still modify the position of the monsec/sttime/coord/fps widgets in compact and distributed Crispy HUD mode.

It's a bit inconsistent with Boom mode, but I don't want to nitpick so much 😆 Overall, I think it's more than enough.

**Comment by fabiangreffrath** [comment]
> It's a bit inconsistent with Boom mode, but I don't want to nitpick so much laughing Overall, I think it's more than enough.

There were still some comments regarding the menu switches, how they should be called and which are missing or probably misleading. I'm not sure what *exactly* is left to do here, though.

**Review by rfomin — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
Wow, that was a journey. 😁

**Pull request merged**